go-controller/pkg/cluster/master.go

```diff
@@ -14,8 +12,171 @@ import (
 
 	"github.com/openshift/origin/pkg/util/netutils"
+	"github.com/sirupsen/logrus"
 )
 
+// RebuildClusterNodes rebuilds db for cluster nodes
+func (cluster *OvnClusterController) RebuildClusterNodes(
+	masterNodeName string, oc *ovn.Controller) error {
+	logrus.Debugf("Rebuild OVN database for cluser nodes")
+	var err error
+
+	// If OvnHA options is enabled, make sure default namespace has the
+	// annotation of current cluster master node's overlay IP address
+	logrus.Debugf("cluster.OvnHA: %t", cluster.OvnHA)
+	if cluster.OvnHA && !cluster.OvnDbExist {
+		logrus.Debugf("HA is enabled and DB doesn't exist!")
+		// Rebuild OVN DB for the k8s nodes
+		err = cluster.UpdateKubeNodes(masterNodeName)
+		if err != nil {
+			return err
+		}
+		// Rebuild OVN DB for the k8s namespaces and all the resource
+		// objects inside the namespace including pods and network
+		// policies
+		err = cluster.UpdateKubeNsObjects(oc)
+		if err != nil {
+			return err
+		}
+		// Update master node IP annotation on default namespace
+		err = cluster.UpdateMasterNodeIP(masterNodeName)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// UpdateKubeNodes rebuilds ovn db for k8s nodes
+func (cluster *OvnClusterController) UpdateKubeNodes(
+	masterNodeName string) error {
+	nodes, err := cluster.Kube.GetNodes()
+	if err != nil {
+		logrus.Errorf("Failed to obtain k8s nodes: %v", err)
+		return err
+	}
+	for _, node := range nodes.Items {
+		subnet, ok := node.Annotations[OvnHostSubnet]
+		if ok {
+			// Create a logical switch for the node and connect it to
+			// the distributed router
+			logrus.Debugf("ovn_host_subnet: %s", subnet)
+			ip, localNet, err := net.ParseCIDR(subnet)
+			if err != nil {
+				return fmt.Errorf("Failed to parse subnet %v: %v", subnet,
+					err)
+			}
+			ip = util.NextIP(ip)
+			n, _ := localNet.Mask.Size()
+			routerIPMask := fmt.Sprintf("%s/%d", ip.String(), n)
+			stdout, stderr, err := util.RunOVNNbctl("--may-exist",
+				"ls-add", node.Name, "--", "set", "logical_switch",
+				node.Name, fmt.Sprintf("other-config:subnet=%s", subnet),
+				fmt.Sprintf("external-ids:gateway_ip=%s", routerIPMask))
+			if err != nil {
+				logrus.Errorf("Failed to create logical switch for "+
+					"node %s, stdout: %q, stderr: %q, error: %v",
+					node.Name, stdout, stderr, err)
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+// UpdateKubeNsObjects rebuilds ovn db for k8s namespaces
+// and pods/networkpolicies inside the namespaces.
+func (cluster *OvnClusterController) UpdateKubeNsObjects(
+	oc *ovn.Controller) error {
+	namespaces, err := cluster.Kube.GetNamespaces()
+	if err != nil {
+		logrus.Errorf("Failed to get k8s namespaces: %v", err)
+		return err
+	}
+	for _, ns := range namespaces.Items {
+		oc.AddNamespace(&ns)
+		pods, err := cluster.Kube.GetPods(ns.Name)
+		if err != nil {
+			logrus.Errorf("Failed to get k8s pods: %v", err)
+			return err
+		}
+		for _, pod := range pods.Items {
+			oc.AddLogicalPortWithIP(&pod)
+		}
```

Comment:
I think we should also handle endpoints.

Reply:
done